### PR TITLE
[MU4] Scores uploading

### PR DESCRIPTION
**Pull request by RomanPudashkin**

### CODE REVIEW

File: src/cloud/internal/cloudservice.cpp [hunk 1]
```diff
@@ -41,6 +42,17 @@ static const QString REFRESH_TOKEN_KEY("refresh_token");
 static const QString DEVICE_ID_KEY("device_id");
 
 constexpr int USER_UNAUTHORIZED_ERR_CODE = 401;
+constexpr int INVALID_SCORE_ID = 0;
+
+int scoreIdFromSourceUrl(const QUrl& sourceUrl)
```
Comment (igorkorsukov) [inline src/cloud/internal/cloudservice.cpp:47]: Better add `static`, else this function will be global

File: src/cloud/internal/cloudservice.cpp [hunk 2]
```diff
@@ -165,7 +169,21 @@ void CloudService::onUserAuthorized()
     m_refreshToken = m_oauth2->refreshToken();
 
     saveTokens();
-    downloadUserInfo();
+
+    if (downloadUserInfo() == RequestStatus::Ok) {
+        m_onUserAuthorizedCallback();
```
Comment (igorkorsukov) [inline src/cloud/internal/cloudservice.cpp:174]: I believe need to add `IF_ASSERT_FAILED(m_onUserAuthorizedCallback)`

File: src/cloud/internal/cloudservice.h [hunk 3]
```diff
@@ -55,7 +57,10 @@ class CloudService : public QObject, public IAuthorizationService, public IUploa
     ValCh<bool> userAuthorized() const override;
     ValCh<AccountInfo> accountInfo() const override;
 
-    framework::ProgressChannel uploadScore(const QByteArray& scoreData, const std::string& title, const QUrl& sourceUrl = QUrl()) override;
+    void uploadScore(system::IODevice& scoreSourceDevice, const QString& title, const QUrl& sourceUrl = QUrl()) override;
+
+    async::Channel<QUrl> sourceUrlReceived() const override;
+    framework::ProgressChannel progressChannel() const override;
```
Comment (igorkorsukov) [inline src/cloud/internal/cloudservice.h:63]: It's strange that `ProgressChannel` in the namespace is a framework, not an namespace async  :)

File: src/cloud/internal/cloudservice.h [hunk 4]
```diff
@@ -76,8 +88,10 @@ private slots:
     network::RequestHeaders headers() const;
 
     RequestStatus downloadUserInfo();
+    RequestStatus doUploadScore(system::IODevice& scoreSourceDevice, const QString& title, const QUrl& sourceUrl = QUrl());
```
Comment (igorkorsukov) [inline src/cloud/internal/cloudservice.h:91]: It is strange that we have `IODevice` in the system namespace, not io.

File: src/cloud/internal/cloudservice.h [hunk 5]
```diff
@@ -76,8 +88,10 @@ private slots:
...
 
-    void setAccountInfo(const AccountInfo& info);
+    using RequestCallback = std::function<RequestStatus()>;
+    void executeRequest(const RequestCallback& requestCallback);
```
Comment (igorkorsukov) [inline src/cloud/internal/cloudservice.h:94]: I think it better to use async::Promise

File: src/cloud/internal/cloudservice.cpp (hunk 2 above)
Comment (RomanPudashkin) [reply src/cloud/internal/cloudservice.cpp:174]: m_onUserAuthorizedCallback can be invalid (see CloudService::signIn()). This callback is needed when you need something to do right after authorization (for example, uploading of the score will continue if a user wasn't authorized when they pressed on "Upload to MuseScore.com")

Comment (igorkorsukov) [reply src/cloud/internal/cloudservice.cpp:174]: Then you just need to check `if (m_onUserAuthorizedCallback)`, otherwise it will crash :)

Comment (RomanPudashkin) [reply src/cloud/internal/cloudservice.cpp:174]: fixed

File: src/cloud/internal/cloudservice.cpp (hunk 1 above)
Comment (RomanPudashkin) [reply src/cloud/internal/cloudservice.cpp]: fixed

File: src/cloud/internal/cloudservice.h (hunk 3 above)
Comment (RomanPudashkin) [reply src/cloud/internal/cloudservice.h:63]: Progress is stored in mu::global. ProgressChannel just an alias. We can put ProgressChannel into mu::async, but it seems a little strange for me

File: src/cloud/internal/cloudservice.h (hunk 4 above)
Comment (RomanPudashkin) [reply src/cloud/internal/cloudservice.h]: moved IODevice into mu::io

File: src/cloud/internal/cloudservice.h [hunk 6]
```diff
@@ -66,6 +71,13 @@ private slots:
     bool updateTokens();
     void clearTokens();
 
+    using OnUserAuthorizedCallback = std::function<void ()>;
+    void authorize(const OnUserAuthorizedCallback& onUserAuthorizedCallback = OnUserAuthorizedCallback());
```
[the remark this answers is not in the document]
Comment (RomanPudashkin) [reply src/cloud/internal/cloudservice.h:75]: Discussed this issue with Igor, decided to leave it as it is

**Pull request merged**